src/base_media_func.h

```diff
@@ -159,9 +159,9 @@ bool BaseMedia<Tbase_set>::AddFile(const char *filename, size_t basepath_length,
 
 	Tbase_set *set = new Tbase_set();
 	IniFile *ini = new IniFile();
-	ini->LoadFromDisk(filename, BASESET_DIR);
-
 	char *path = stredup(filename + basepath_length);
+	ini->LoadFromDisk(path, BASESET_DIR);
```

Comment:
I *think* the old place was correct, but it should be NO_DIRECTORY as second argument.  But .. this code is difficult to follow, and it seems a hack on a hack.

Reply:
From what I can see, ScanPath is ran on BASESET_DIR. Every hit it calls AddFile, with 'filename' the full path, and 'basepath_length' with the offset of the searchpath found. Next LoadFromDisk is called on the filename, and using BASESET_DIR here causes it to search the searchpaths again. But AddFile already knows where the hit is. At least .. this is my (limited) understanding :D

Reply:
Another look might suggest that my suggestion kills TAR support. Not sure. Code is weird.

Reply:
Using `NO_DIRECTORY` there breaks another file discovery:
```[5528] dbg: [console] /home/<user>/source/src/fileio.cpp:1391 ScanPath Found file /home/<user>/.openttd/content_download/baseset/OpenGFX-0.5.2.tar
[5528] dbg: [console] /home/<user>/source/src/fileio.cpp:1378 ScanPath Valid file: /home/<user>/.openttd/content_download/baseset/
[5528] dbg: [grf] /home/<user>/source/src/base_media_func.h:158 AddFile Checking opengfx-0.5.2/opengfx.obg for base graphics set
[5528] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load opengfx-0.5.2/opengfx.obg
[5528] dbg: [console] /home/<user>/source/src/base_media_func.h:165 AddFile path = opengfx-0.5.2/opengfx.obg
[5528] dbg: [console] /home/<user>/source/src/base_media_func.h:172 AddFile path = opengfx-0.5.2/
[5528] dbg: [grf] /home/<user>/source/src/base_media_func.h:46 FillSetDetails Base graphicsset detail loading: name field missing.
[5528] dbg: [grf] /home/<user>/source/src/base_media_func.h:46 FillSetDetails   Is opengfx-0.5.2/opengfx.obg readable for the user running OpenTTD?
```
Normal behaviour (trunk) is:
```[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:1391 ScanPath Found file /home/<user>/.openttd/content_download/baseset/OpenGFX-0.5.2.tar
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:1378 ScanPath Valid file: /home/<user>/.openttd/content_download/baseset/
[5781] dbg: [grf] /home/<user>/source/src/base_media_func.h:158 AddFile Checking opengfx-0.5.2/opengfx.obg for base graphics set
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load /home/<user>/source/baseset/opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load /home/<user>/.openttd/baseset/opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load /home/<user>/source/bin/baseset/opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load /usr/local/share/games/openttd/baseset/opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/fileio.cpp:421 FioFOpenFileSp Trying to load /home/<user>/.openttd/content_download/baseset/opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/base_media_func.h:165 AddFile path = opengfx-0.5.2/opengfx.obg
[5781] dbg: [console] /home/<user>/source/src/base_media_func.h:172 AddFile path = opengfx-0.5.2/
```
What's bothering with that `AddFile` behaviour is that, even though a specific file is found in `ScanPath`, it triggers yet another file discovery in *all* of the basesets directories instead of merely loading from a subdirectory of the found file (for cases like this one where the file is an archive, seen as a subdirectory by the loading system).

However, all that seems to go beyong that quick & dirty fix, calling for a global rewrite...

Reply:
`IniLoadFile::LoadFromDisk` then calls `IniFile::OpenFile`, which is a mere proxy to `FioFOpenFile`.
This function seems thus to be central in all file loading operations.

While this function handles cases were full paths are provided to it, as I pointed out it was working before (look towards its end), it's definitely inefficient to do so.

Might even be a god idea to rewrite this function to that full paths are detected early in it, and rather than blindly attempting to load resources, check of the beginning of a filename matches any of the registered search paths, and then directly jump to fullpath loading attempt?

And of course, I am not even talking about the quality of the code in `base_media_func.h`, which is more of your league.